File: examples/java/src/main/java/io/superdurable/dex/primitives/clientapis/ClientApisFlow.java
/*
 * Copyright (c) 2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.superdurable.dex.primitives.clientapis;

import io.superdurable.dex.Attribute;
import io.superdurable.dex.AttributeIndex;
import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.PersistenceSchema;
import io.superdurable.dex.Step;
import io.superdurable.dex.StepDecision;
import io.superdurable.dex.StepList;
import org.springframework.stereotype.Component;

@Component
public final class ClientApisFlow implements Flow<String> {
    static final String KEYWORD_KEY = "CustomKeywordField";

    public final Attribute<String> keyword = Attribute.define(
            KEYWORD_KEY,
            String.class,
            new AttributeIndex(AttributeIndex.Type.KEYWORD));
    private final ClientApisStep start = new ClientApisStep();

    @Override
    public StepList<String> getSteps() {
        return StepList.startStep(start);
    }

    @Override
    public PersistenceSchema getPersistenceSchema() {
        return PersistenceSchema.of(keyword);
    }

    final class ClientApisStep implements Step<String> {
        @Override
        public Class<String> getInputType() {
            return String.class;
        }

        @Override
        public StepDecision execute(final Context context, final String input) {
            keyword.set(context, input);
            return StepDecision.gracefulComplete(input);
        }
    }
}

File: examples/typescript/src/primitives/client-apis/client-apis-flow.ts
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  Attribute,
  IndexType,
  StepList,
  gracefulComplete,
  stringCodec,
  type Context,
  type Flow,
  type PersistenceSchema,
  type Step,
  type StepDecision,
} from "@superdurable/dex";

export const KEYWORD_KEY = "CustomKeywordField";

class ClientApisStep implements Step<string> {
  public readonly inputCodec = stringCodec;

  public constructor(private readonly flow: ClientApisFlow) {}

  public getStepType(): string {
    return "ClientApisStep";
  }

  public execute(context: Context, input: string): StepDecision {
    this.flow.keyword.set(context, input);
    return gracefulComplete(input);
  }
}

export class ClientApisFlow implements Flow<string> {
  public readonly keyword = new Attribute(KEYWORD_KEY, stringCodec, {
    type: IndexType.KEYWORD,
  });
  private readonly index = new ClientApisStep(this);

  public getFlowType(): string {
    return "ClientApisFlow";
  }

  public getSteps() {
    return StepList.startStep(this.index);
  }

  public getPersistenceSchema(): PersistenceSchema {
    return { attributes: [this.keyword] };
  }
}

export const clientApisFlow = new ClientApisFlow();
